docs(SafeArea): SafeArea ScrollViewer Documentation (#376)
unoplatform/uno.toolkit.ui: doc/controls/SafeArea.md

The `ApplicationView.VisibleBounds` is the rectangular area of the screen which is completely unobscured by any window decoration, such as the status bar, rounded screen corners, or any type of screen notch.

`SafeArea` can also be used to specify certain areas of the UI that should adapt its layout in order to avoid being covered by any sort of soft-input panel, such as the on-screen keyboard. This is done by observing the state of the keyboard and treating the area that it occupies when open as part of the "unsafe" area of the screen.

In some cases, it is acceptable for visible content to be partially obscured (a page background for example) and it should extend to fill the entire window. Other types of content should be restricted to the visible bounds (for instance: readable text, or interactive controls). `SafeArea` enables this kind of fine-grained control over responsiveness to the safe and "unsafe" areas of the screen.

## Properties

### Remarks

> `SafeArea` can be used as a `Panel` or as a set of attached properties on another `Panel`, much like the `ScrollViewer`:

Reviewer: Sorry, I mean "panel" in the literal sense, like a container. An uwp `Panel` has a different meaning, which typically "_positions and arranges child objects_" (emphasis on the plural part).
Perhaps `ContentControl` is more appropriate in this context, also given that both SV and SA inherit from it.

Reviewer: do a ctrl+f for `panel` and replace them with "control" or "container" whichever fits the best in its context.

```xml
xmlns:utu="using:Uno.Toolkit.UI"

<!-- as attached property on another panel -->
<Grid utu:SafeArea.Insets="Left,Top,Right,Bottom">

>  <!-- Content -->

Reviewer: ```suggestion
    <!-- Content -->
```

</Grid>

> <!-- and/or -->
> 
> <!-- as a Panel -->

Reviewer: ```suggestion
<!-- or, as a Panel -->
```

<SafeArea Insets="Left,Top,Right,Bottom">

>  <!-- Content -->

Reviewer: ```suggestion
    <!-- Content -->
```

</SafeArea>
```

> [!WARNING]

> > In most cases, the attached properties and the `SafeArea` `Panel` can be used interchangeably. However, be aware that using `SafeArea` as a `Panel` while `Insets` contains `InsetMask.SoftInput` will introduce a `ScrollViewer` into the visual tree as the content root of the `SafeArea`. Please refer to the [SoftInput usage section](#using-insetmasksoftinput-for-on-screen-keyboards).

Reviewer: ```suggestion
> In most cases, the attached properties and the `SafeArea` control can be used interchangeably. However, be aware that using `SafeArea` as a control while `Insets` contains `InsetMask.SoftInput` will introduce a `ScrollViewer` into the visual tree as the content root of the `SafeArea`. Please refer to the [SoftInput usage section](#using-insetmasksoftinput-for-on-screen-keyboards).
```

> The `InsetMask` enum can represent a single bound or it can be combined with multiple values as shown in the above example XAML. `InsetMask` has the following available values:

Reviewer: ```suggestion
The `InsetMask` enum can represent a single edge/side or it can be composed of multiple values (eg: `InsetMask="Left, Right"`). `InsetMask` has the following available values:
```

Attempting to use `SafeArea` to adapt to the keyboard without including a `ScrollViewer` may not always yield the expected visual result. The behavior of a keyboard-aware `SafeArea` depends on the structure of the layout. For example, a simple login page like the one below can benefit from `SafeArea` to ensure that the Username/Password fields, when in focus, are kept visible above the keyboard.

#### Sample Login Page

In this initial example, notice that 

<div style="display:flex;flex-direction:row;flex-wrap:wrap">
  <div style="flex-shrink:1">
    <img width="350px" src="../assets/safearea-login-static.gif" />
  </div>
  <div style="flex:2">

```xml
<Page ...
	  xmlns:utu="using:Uno.Toolkit.UI">
	<Grid Padding="50,0">
		<Grid.RowDefinitions>
			<RowDefinition Height="Auto" />
			<!--  Spacer  -->
			<RowDefinition Height="40" />
			<RowDefinition Height="*" />
		</Grid.RowDefinitions>
		<Image Source="ms-appx:///Assets/uno-logo.png"
				VerticalAlignment="Stretch"
				HorizontalAlignment="Stretch" />

> 		<StackPanel Grid.Row="2"

Reviewer: would be nice to explain a bit why `Logo` is always outside of SafeArea, while `Header` is
or how they will behave differently

> 		<StackPanel Grid.Row="2"

Author: Well here's what I'm struggling with: should the "proper" usage xaml example be:

**SafeArea**
    Grid
        Image
        _Spacer_
        StackPanel

Or

Grid
    Image
    _Spacer_
    **SafeArea**
        StackPanel

Both using SafeArea as a Panel for the ScrollViewer

It looks weird if the whole page isn't wrapped in a ScrollViewer, like you just get the form being clipped as it reaches the bottom of the logo

		</StackPanel>
	</Grid>
</Page>
```
  </div>
</div>

# SafeAreaScroll
```xml
	<utu:SafeArea Insets="SoftInput">
		<Grid Padding="50,0">
			<Grid.RowDefinitions>
				<RowDefinition Height="Auto" />
				<!--  Spacer  -->
				<RowDefinition Height="40" />
				<RowDefinition Height="*" />
			</Grid.RowDefinitions>
			<Image Source="ms-appx:///Assets/uno-logo.png"
				   VerticalAlignment="Stretch"
				   HorizontalAlignment="Stretch" />

			<StackPanel Grid.Row="2"
						Spacing="20">
				<TextBlock Text="Welcome to Uno!"
						   HorizontalAlignment="Center"
						   FontSize="35" />
				<PersonPicture ProfilePicture="ms-appx:///Assets/profile.png" />
				<TextBox x:Name="TextBox"
						 PlaceholderText="Username" />
				<PasswordBox x:Name="PasswordBox"
							 PlaceholderText="Password" />
				<StackPanel Orientation="Horizontal"
							Spacing="4">
					<CheckBox Padding="0"
							  MinWidth="0" />
					<TextBlock Text="Remember me"
							   VerticalAlignment="Center" />
				</StackPanel>
				<Button Content="Login"
						x:Name="LoginBtn"
						VerticalAlignment="Top"
						HorizontalAlignment="Stretch"
						Margin="0,30" />
			</StackPanel>
		</Grid>
	</utu:SafeArea>
```

> # safe area attached

Author: Yeah still organizing everything into sections now

>
> - Attach `SafeArea` properties to an existing `ScrollViewer` that is wrapping the content that you would like to adapt to the on-screen keyboard.
> - Or, use `SafeArea` as a `Panel` and have it wrap the relevant content. The `Panel` will automatically include a `ScrollViewer` that wraps its content as long as the `Insets` property contains `InsetMask.SoftInput`.

Attempting to use `SafeArea` to adapt to the keyboard without including a `ScrollViewer` may not always yield the expected visual result. The behavior of a keyboard-aware `SafeArea` depends on the structure of the layout. For example, a simple login page like the one below can benefit from `SafeArea` to ensure that the Username/Password fields, when in focus, are kept visible above the keyboard.

#### Sample Login Page

Notice in this first example (without `SafeArea` in use) that the Username and Password field are covered by the keyboard and the UI above the keyboard is not scrollable so the relevant views cannot be brought into the visible frame.

<table>
  <tr>
    <th>Page</th>
    <th>XAML</th>
  </tr>
  <tr>
   <td><img src="../assets/safearea-login-static.gif" width="400px"/> </td>
    <td>

```xml

> <Page ...
>    xmlns:utu="using:Uno.Toolkit.UI">
>     <Grid Padding="50,0">
>         <Grid.RowDefinitions>
>             <!--  0: Logo, 1: Spacer, 2: FormPanel  -->
>             <RowDefinition Height="Auto" />
>             <RowDefinition Height="40" />
>             <RowDefinition Height="*" />
>         </Grid.RowDefinitions>
>         <!-- Logo -->
>         <Image Source="ms-appx:///Assets/uno-logo.png"
>                 VerticalAlignment="Stretch"
>                 HorizontalAlignment="Stretch" />
>         <!-- FormPanel -->
>         <StackPanel Grid.Row="2"
>                     x:Name="FormPanel"
>                     Spacing="20">
>             <TextBlock Text="Welcome to Uno!"
>                         HorizontalAlignment="Center"
>                         FontSize="35" />
>             <PersonPicture ProfilePicture="ms-appx:///Assets/profile.png" />
>             <TextBox x:Name="TextBox"
>                         PlaceholderText="Username" />
>             <PasswordBox x:Name="PasswordBox"
>                             PlaceholderText="Password" />
>             <StackPanel Orientation="Horizontal"
>                         Spacing="4">
>                 <CheckBox Padding="0"
>                             MinWidth="0" />
>                 <TextBlock Text="Remember me"
>                             VerticalAlignment="Center" />
>             </StackPanel>
>             <Button Content="Login"
>                     x:Name="LoginBtn"
>                     VerticalAlignment="Top"
>                     HorizontalAlignment="Stretch"
>                     Margin="0,30" />
>         </StackPanel>
>     </Grid>
> </Page>

Reviewer: v-align attributes:
```suggestion
<Page xmlns:utu="using:Uno.Toolkit.UI" ...>
    <Grid Padding="50,0">
        <Grid.RowDefinitions>
            <!-- 0: Logo, 1: Spacing, 2: FormPanel -->
            <RowDefinition Height="Auto" />
            <RowDefinition Height="40" />
            <RowDefinition Height="*" />
        </Grid.RowDefinitions>
        <!-- Logo -->
        <Image Source="ms-appx:///Assets/uno-logo.png"
               VerticalAlignment="Stretch"
               HorizontalAlignment="Stretch" />
        <!-- FormPanel -->
        <StackPanel Grid.Row="2"
                    x:Name="FormPanel"
                    Spacing="20">
            <TextBlock Text="Welcome to Uno!"
                       HorizontalAlignment="Center"
                       FontSize="35" />
            <PersonPicture ProfilePicture="ms-appx:///Assets/profile.png" />
            <TextBox x:Name="TextBox"
                     PlaceholderText="Username" />
            <PasswordBox x:Name="PasswordBox"
                         PlaceholderText="Password" />
            <StackPanel Orientation="Horizontal"
                        Spacing="4">
                <CheckBox Padding="0"
                          MinWidth="0" />
                <TextBlock Text="Remember me"
                           VerticalAlignment="Center" />
            </StackPanel>
            <Button Content="Login"
                    x:Name="LoginBtn"
                    VerticalAlignment="Top"
                    HorizontalAlignment="Stretch"
                    Margin="0,30" />
        </StackPanel>
    </Grid>
</Page>
```

```

</td>
  </tr>
</table>

In this next example, we attempt to have the UI adapt to the keyboard by attaching the `SafeArea.Insets` property to the `StackPanel` that contains the login form. We can see that there is no visual change when compared to the previous example where `SafeArea` was not present. This is due to the fact that there are "hard" constraints within the XAML, such as the hardcoded Spacer Grid Row and the `Auto` Row containing Uno logo image.

<table>
  <tr>
    <th>Page</th>
    <th>XAML</th>
  </tr>
  <tr>
   <td><img src="../assets/safearea-login-static.gif" width="400px"/> </td>
    <td>

```diff

>  ...
> 
>   <StackPanel Grid.Row="2"
>             x:Name="FormPanel"
> +           utu:SafeArea.Insets="SoftInput"
>             Spacing="20">
>     <TextBlock Text="Welcome to Uno!"
>                 HorizontalAlignment="Center"
>     FontSize="35" />
>     <PersonPicture ProfilePicture="ms-appx:///Assets/profile.png" />
>     <TextBox x:Name="TextBox"
>                 PlaceholderText="Username" />
>     <PasswordBox x:Name="PasswordBox"
>                     PlaceholderText="Password" />
>     <StackPanel Orientation="Horizontal"
>                 Spacing="4">
>         <CheckBox Padding="0"
>                     MinWidth="0" />
>         <TextBlock Text="Remember me"
>                     VerticalAlignment="Center" />
>     </StackPanel>
>     <Button Content="Login"
>             x:Name="LoginBtn"
>             VerticalAlignment="Top"
>             HorizontalAlignment="Stretch"
>             Margin="0,30" />
> </StackPanel>
> 
>  ...

Reviewer: ```suggestion
 <StackPanel Grid.Row="2"
             x:Name="FormPanel"
+            utu:SafeArea.Insets="SoftInput"
             Spacing="20">
     <TextBlock Text="Welcome to Uno!"
                HorizontalAlignment="Center"
                FontSize="35" />
     <PersonPicture ProfilePicture="ms-appx:///Assets/profile.png" />
     <TextBox x:Name="TextBox"
              PlaceholderText="Username" />
     <PasswordBox x:Name="PasswordBox"
                  PlaceholderText="Password" />
     <StackPanel Orientation="Horizontal"
                 Spacing="4">
         <CheckBox Padding="0"
                   MinWidth="0" />
         <TextBlock Text="Remember me"
                    VerticalAlignment="Center" />
     </StackPanel>
     <Button Content="Login"
             x:Name="LoginBtn"
             VerticalAlignment="Top"
             HorizontalAlignment="Stretch"
             Margin="0,30" />
 </StackPanel>
```

```

</td>
  </tr>
</table>

The recommended solution in this case would be to use a combination of `SafeArea` and a `ScrollViewer` that will wrap your `Page` content. When using `SafeArea` as a `Panel`, a `ScrollViewer` is automatically included as the root element of `SafeArea` as long as the `Insets` property contains the `InsetMask.SoftInput` value.

<table>
  <tr>
    <th>Page</th>
    <th>XAML</th>
  </tr>
  <tr>
   <td><img src="../assets/safearea-login-scroll.gif" width="400px"/> </td>
    <td>

```diff

> <Page ...
>     xmlns:utu="using:Uno.Toolkit.UI">
>     ...
> +    <utu:SafeArea Insets="SoftInput">
>         <Grid Padding="50,0">
>             <Grid.RowDefinitions>
>                 <!--  0: Logo, 1: Spacer, 2: FormPanel  -->
>                 <RowDefinition Height="Auto" />
>                 <RowDefinition Height="40" />
>                 <RowDefinition Height="*" />
>             </Grid.RowDefinitions>
>             ...
>         </Grid>
> +    </utu:SafeArea>
>     ...
> </Page>

Reviewer: ```suggestion
<Page xmlns:utu="using:Uno.Toolkit.UI" ...>
+   <utu:SafeArea Insets="SoftInput">
        <Grid Padding="50,0">
            <Grid.RowDefinitions>
                <!--  0: Logo, 1: Spacer, 2: FormPanel  -->
                <RowDefinition Height="Auto" />
                <RowDefinition Height="40" />
                <RowDefinition Height="*" />
            </Grid.RowDefinitions>

            <!-- logo and form... -->
        </Grid>
+   </utu:SafeArea>
</Page>
```

Reviewer: same for tables below

                <PasswordBox x:Name="PasswordBox"
                                PlaceholderText="Password" />
                <StackPanel Orientation="Horizontal"
                            Spacing="4">
                    <CheckBox Padding="0"
                                MinWidth="0" />
                    <TextBlock Text="Remember me"
                                VerticalAlignment="Center" />
                </StackPanel>
                <Button Content="Login"
                        x:Name="LoginBtn"
                        VerticalAlignment="Top"
                        HorizontalAlignment="Stretch"
                        Margin="0,30" />
            </StackPanel>
    +    </utu:SafeArea>
        ...
    </Page>
    ```

    </td>
    </tr>
    </table>

## Notes

- `SafeArea` is able to adapt to views that are only _partially_ obscured by applying the minimum amount of `Padding`/`Margin` needed until the content is fully inside the visible bounds.
- When a control already has a non-zero `Padding`/`Margin`, `SafeArea` takes those values into consideration when calculating the minimum amount of pixels needed for the view to be within the safe area.

> - `SafeArea` on WinAppSDK/Desktop does not have any effect. It is present to allow for same-XAML across platforms.

Reviewer: ```suggestion
- `SafeArea` on WinAppSDK/Desktop currently does not have any effect. It is present to allow for same XAML across platforms.
```

Reviewer: this doesnt work on uwp on windows and winui on windows? or just the latter?

Author: Doesn't work on WinUI Windows
